### PR TITLE
Add timeout support to GraphQL executions

### PR DESCRIPTION
**Pull request by Shane32**

### CODE REVIEW

File: src/GraphQL.Tests/Execution/DocumentExecuterTests.cs [hunk 1]
```diff
@@ -327,6 +328,122 @@ public async Task Execution_Wraps_OtherExeceptions(bool throwOnUnhandledExceptio
...
+    [InlineData(false, false)]
+    public async Task Timeout_Infinite_PassThrough(bool rethrowTimeout, bool withTimeout)
+    {
+        using var cts = new CancellationTokenSource();
+        var query = new ObjectGraphType();
+        query.Field<StringGraphType>("test")
+            .ResolveAsync(async context =>
+            {
+                context.CancellationToken.CanBeCanceled.ShouldBeTrue();
+                cts.CancelAfter(100); // ensure asynchronous cancellation by using CancelAfter
```
Comment (gao-artur) [inline src/GraphQL.Tests/Execution/DocumentExecuterTests.cs:427]: Starting .NET 8 there is a `CancelAsync`. Not sure if it's Pollyfilled
https://learn.microsoft.com/en-us/dotnet/api/system.threading.cancellationtokensource.cancelasync?view=net-8.0

File: src/GraphQL/Execution/DocumentExecuter.cs [hunk 2]
```diff
@@ -213,10 +225,23 @@ await listener.AfterExecutionAsync(context)
...
+            // Re-throw the original cancellation exception when the the original CancellationToken is canceled
+            // (e.g. the client disconnects)
             throw;
         }
+        catch (OperationCanceledException) when (options.CancellationToken.IsCancellationRequested)
+        {
+            // If the operation was canceled due to a timeout, return a result with no data and an error
+            // or throw a TimeoutException based on the configuration
+            if (options.TimeoutAction == TimeoutAction.ThrowTimeoutException)
+                throw new GraphQLTimeoutException();
```
Comment (gao-artur) [inline src/GraphQL/Execution/DocumentExecuter.cs:239]: Should we pass the original exception as an inner exception?

**Review by gao-artur — APPROVED** [review]

File: src/GraphQL.Tests/Execution/DocumentExecuterTests.cs (hunk 1 above)
Comment (Shane32) [reply src/GraphQL.Tests/Execution/DocumentExecuterTests.cs:427]: Nice!

File: src/GraphQL/Execution/DocumentExecuter.cs (hunk 2 above)
Comment (Shane32) [reply src/GraphQL/Execution/DocumentExecuter.cs]: Done

**Pull request merged**